docs/content/sips/007-deployment-auth.md

```diff
@@ -0,0 +1,158 @@
+
+## Error flows
+
+### Auth failure on deploy
+
+Deployment would fail if the token was incorrect or expired.
+
+> QUESTION: In this case should we simply fail, or should we attempt login?  _Note:_ if we decide to attempt login we _must_ provide a flag to suppress this (for programmatic consumers who have their own login flows).
+
+It would be useful to have an unambiguous way of reporting this for programmatic consumers (e.g. distinguished error code, JSON output option), so they could display login UI or provide a specific prompt.  That may be a bit much to manage in the first drop though.
+
+## `spin login` detailed considerations
+
+`spin login` is a new command and a relatively complex user experience.  This section describes the behaviour in more detail.
+
+### Browser interaction
+
+As part of device auth flow, the user needs to go to the verification URL in their browser and paste in the code.
+
+> QUESTION: Should we attempt to open the browser automatically?
+> * Pro: it's more convenient
+> * Con: it may be unexpected (we have not asked for consent or warned it will happen)
+> * Con: it may work differently across browsers and operating systems, which may require explanation in the documentation
+
+> QUESTION: Should we copy the user code to the clipboard?
+> * Pro: it's convenient
+> * Con: the user may already have stuff on their clipboard
+> * Con: again, documentation needs to consider differences across OSes
+
+(For both of these questions, is there prior art to which we can refer?)
+
+### Verification polling
+
+The device auth flow involves the client polling the server using the device code as the auth token. Each poll can have three outcomes:
+
+* 200 and the response contains a token: user has authorised this device
+* 200 and the response doesn't contain a token: user has not yet authorised this device
+* Server error
+* Network error
+
+We propose to poll for two minutes. If the user has not entered the code by then, `spin login` fails and exits.
+
+Errors do _not_ cause polling to terminate. We do not want a transient network error or server hiccup to block login. That said, if we can identify server responses that are necessarily fatal, we can in future terminate early on those.
+
+> QUESTION: Are there any accessibility considerations around the timeout? For example, should we provide a way to override it (e.g. longer timeout, poll only when the user is ready) for users who might not be able to execute the process within two minutes? Is there prior art from other applications we can refer to here?
+
+### Token storage
+
+When the user logs into a server, we need to store:
+
+* The server URL
+* Any server connection options, such as the `--insecure` flag
+* The token
+
+A simple solution is to store these in a file under a non-roaming directory. This is not fantastically secure for tokens, but may be enough for the preview.
```

Comment:
what does "non-roaming" mean?

Reply:
> This is not fantastically secure for tokens

Yeah. This is something we need to fix on the server side via refresh tokens.

In the current system, we only issue an auth token. If an attacker manages to get a hold of that auth token, they can use it for as long as it's valid (currently 7 days from date of issue). Even if you change your password. That’s because the server will trust the token if its signature is valid. The only way to invalidate the token is to change the key that was used to sign it, and that has the consequence of invalidating everyone else’s tokens.

With refresh tokens, you can send both the auth token as well as the refresh token to fetch a fresh auth token. This means the user will not need to re-enter their username/password or re-enter a device code when it expires, and the server can shorten the time window that the auth token is valid down to a more reasonable limit (generally between 1-8 hours). If the auth token is compromised, a server administrator can invalidate the refresh token, forcing the user to log back in. If a user does not use the app within the time limit of the refresh token (usually between 7 days), then they are forced to log back in.

Reply:
A good crash course on auth vs refresh tokens can be found here: https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/

Reply:
"Non-roaming" means it lives only on this machine and is not synced to other machines.  (As I assume a "device code" works only on the machine it came from.)  E.g. on Windows the %USERPROFILE%/AppData/Roaming directory follows the user around but %USERPROFILE%/AppData/Local is specific to the user's account _on that machine_.  Not sure of analogues on other OSes.

Reply:
> Not sure of analogues on other OSes.

https://docs.rs/dirs/4.0.0/dirs/fn.data_local_dir.html